src/main/java/com/example/SOPSbackend/security/BasicUserDetails.java

```diff
@@ -0,0 +1,101 @@
+package com.example.SOPSbackend.security;
+
+import com.example.SOPSbackend.model.Admin;
+import com.example.SOPSbackend.model.BasicUser;
+import com.example.SOPSbackend.model.Doctor;
+import com.example.SOPSbackend.model.Patient;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+
+public class BasicUserDetails implements UserDetails {
+    private final String email;
+    private final String password;
+    private final Role role;
+
+    private final Object databaseObject;
+
+    public BasicUserDetails(BasicUser user) {
+        email = user.getEmail();
+        password = user.getPassword();
+        role = user.getRole();
+        databaseObject = user;
+    }
+
+    public Role getRole() {
+        return role;
+    }
+
+    public boolean isAdmin() {
+        return role == Role.ADMIN;
+    }
+
+    public boolean isDoctor() {
+        return role == Role.DOCTOR;
+    }
+
+    public boolean isPatient() {
+        return role == Role.PATIENT;
+    }
+
+    public Admin asAdmin() {
```

Comment:
hmm, co tu się w ogóle dzieje - ta metoda jest wywoływana w miejscu, gdzie oczekujemy BasicUser. Dlaczego w ten dziwny sposób castujemy ją na lekarza?

Reply:
Analogicznie reszta tych metdod. 

btw 1 instanceof nie jest aż takie złe, nie można od razu tworzyć właściwych obiektów przez jakiś wzorzec projektowy? (jakikolwiek, tutaj najprosciej abstract factory - BasicUser jest naszą abstrakcją - linia 19 BasicUserService). Potem instanceof - skoro juz i tak sprawdzamy typ, to lepiej jest to robic w sposob przejrzysty a nie chowac sie za funkcjami.

btw 2: WAZNE: cos jest mocno zle z hierarchią, że musimy robic takie rzeczy. Pewnie coś przegapiłem, ale tutaj można po prostu zwracać usera z basic user info, a za "key" w linijce 25 SuccessfulAuthReponse.java przyjąć Role#key (dodać pole do Enuma) lub jeżeli chcemy oddzielić warstwy aplikacji to DatabaseHelper#roleToKey(role)

Czyli proponuję połączenie dwóch rozwiązań. Bez tego wygląda to dosyć źle. Oczywiście mogłem coś przegapić. @adamczykt 

IMO

Reply:
Masz całkowitą rację, już to poprawiłem.